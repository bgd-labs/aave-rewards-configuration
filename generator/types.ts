import * as addressBook from '@bgd-labs/aave-address-book';
import {LiquidityMiningSetup, LiquidityMiningUpdate, UmbrellaRewardsUpdate} from './features/types';

export const V3_POOLS = [
  'AaveV3Ethereum',
  'AaveV3EthereumLido',
  'AaveV3EthereumEtherFi',
  'AaveV3Polygon',
  'AaveV3Avalanche',
  'AaveV3Optimism',
  'AaveV3Arbitrum',
  'AaveV3Metis',
  'AaveV3Base',
  'AaveV3BaseSepolia',
  'AaveV3Gnosis',
  'AaveV3Scroll',
  'AaveV3BNB',
  'AaveV3ZkSync',
  'AaveV3Linea',
  'AaveV3Celo',
  'AaveV3Sonic',
  'AaveV3Soneium',
] as const satisfies readonly (keyof typeof addressBook)[];

export const UMBRELLA_POOLS = [
  'UmbrellaEthereum',
  'UmbrellaBaseSepolia',
] as const satisfies readonly (keyof typeof addressBook)[];

export const POOLS = [...V3_POOLS] as const satisfies readonly (keyof typeof addressBook)[];

export type PoolIdentifier = (typeof POOLS)[number];
export type PoolIdentifierV3 = (typeof V3_POOLS)[number];
export type UmbrellaIdentifier = (typeof UMBRELLA_POOLS)[number];

export interface Options {
  force?: boolean;
  feature: FEATURE;
  pool: PoolIdentifier;
  title?: string;
  shortName: string;
  configFile?: string;
  date: string;
}

export type PoolConfigs = Partial<Record<PoolIdentifier, PoolConfig>>;

export type CodeArtifact = {
  code?: {
    constants?: string[];
    fn?: string[];
    execute?: string[];
  };
};

export enum FEATURE {
  UPDATE_UMBRELLA_REWARDS = 'UPDATE_UMBRELLA_REWARDS',
  SETUP_LM = 'SETUP_LM',
  UPDATE_LM = 'UPDATE_LM',
}

export interface FeatureModule<T extends {} = {}> {
  description: string;
  value: FEATURE;
  cli: (args: {options: Options; pool: PoolIdentifier; cache: PoolCache}) => Promise<T>;
  build: (args: {
    options: Options;
    pool: PoolIdentifier;
    cache: PoolCache;
    cfg: T;
  }) => Promise<CodeArtifact>;
}

export type ConfigFile = {
  rootOptions: Options;
  poolOptions: Partial<Record<PoolIdentifier, Omit<PoolConfig, 'artifacts'>>>;
};

export type PoolCache = {blockNumber: number};

export interface PoolConfig {
  artifacts: CodeArtifact[];
  configs: {
    [FEATURE.UPDATE_UMBRELLA_REWARDS]?: UmbrellaRewardsUpdate[];
    [FEATURE.SETUP_LM]?: LiquidityMiningSetup;
    [FEATURE.UPDATE_LM]?: LiquidityMiningUpdate;
  };
  cache: PoolCache;
}

export type Files = {
  jsonConfig: string;
  payloadTest: {pool: PoolIdentifier; payloadTest: string; contractName: string};
};
