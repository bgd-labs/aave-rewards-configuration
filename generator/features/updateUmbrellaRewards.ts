import * as addressBook from '@bgd-labs/aave-address-book';
import {IUmbrellaRewardsController_ABI} from '@bgd-labs/aave-address-book/abis';
import {CHAIN_ID_CLIENT_MAP} from '@bgd-labs/js-utils';
import {select} from '@inquirer/prompts';
import {Hex, getContract} from 'viem';
import {CodeArtifact, FEATURE, FeatureModule} from '../types';
import {UmbrellaRewardsUpdate} from './types';
import {addressPrompt} from '../prompts/addressPrompt';
import {numberPrompt, numberPromptInDays, numberPromptNoTransform} from '../prompts/numberPrompt';
import {
  umbrellaStkAssetsSelectPrompt,
  addressCheckboxPromptWithSymbol,
  translateUmbrellaAssetLibUnderlying,
} from '../prompts/assetsSelectPrompt';
import {
  getTokenDecimals,
  getUmbrellaFromPool,
  getPoolChain,
  CHAIN_TO_CHAIN_ID,
  getTokenSymbols,
  getDefaultCollector,
  getMaxEmissionsPerSecondToReadable,
} from '../common';

export async function fetchUmbrellaRewardsUpdateParams({pool}): Promise<UmbrellaRewardsUpdate[]> {
  const chainId: number = CHAIN_TO_CHAIN_ID[getPoolChain(pool)];
  const umbrella = getUmbrellaFromPool(pool);

  let asset = await umbrellaStkAssetsSelectPrompt({
    message: 'Select the umbrella asset for which rewards needs to be updated:',
    pool,
    required: true,
  });
  let assetAddress: Hex;
  if (asset == 'custom') {
    asset = await addressPrompt({
      message: 'Enter the address of the umbrella asset manually:',
      required: true,
    });
    assetAddress = asset as Hex;
  } else {
    assetAddress = addressBook[umbrella].UMBRELLA_STAKE_ASSETS[asset].STAKE_TOKEN;
  }

  const rewardsControllerContract = getContract({
    abi: IUmbrellaRewardsController_ABI,
    client: CHAIN_ID_CLIENT_MAP[chainId],
    address: addressBook[umbrella].UMBRELLA_REWARDS_CONTROLLER,
  });
  const availableRewardAddresses = await rewardsControllerContract.read.getAllRewards([
    assetAddress,
  ]);
  const symbols = await getTokenSymbols(availableRewardAddresses as Hex[], chainId);

  const rewardsAddresses = await addressCheckboxPromptWithSymbol(
    availableRewardAddresses as Hex[],
    symbols,
    'Select the rewards you wish to update for the asset'
  );
  const input: UmbrellaRewardsUpdate[] = [];

  for (const reward of rewardsAddresses) {
    const currentConfig = await rewardsControllerContract.read.getRewardData([
      assetAddress,
      reward.asset,
    ]);
    console.log('----------------------------------------------------------');
    console.log(
      `Current on-chain configuration for the reward: ${reward.symbol} and asset: ${asset}:`
    );
    console.log(
      `maxEmissionsPerSecond: ${
        currentConfig.maxEmissionPerSecond
      } (${await getMaxEmissionsPerSecondToReadable(
        chainId,
        reward.asset,
        Number(currentConfig.maxEmissionPerSecond)
      )})`
    );
    console.log(
      `distributionEnd: ${currentConfig.distributionEnd} (${new Date(
        Number(currentConfig.distributionEnd) * 1000
      ).toDateString()})`
    );
    console.log('----------------------------------------------------------');

    const maxEmissionsPerSecondChoice = await select({
      message: `Please input the maxEmissionsPerSecond you want to configure for the reward: ${reward.symbol} and asset: ${asset}`,
      choices: [
        {name: 'Keep maxEmissionsPerSecond the same as current', value: 'current'},
        {name: 'Enter maxEmissionsPerSecond in token units / days', value: 'units'},
        {name: 'Enter raw maxEmissionsPerSecond', value: 'raw'},
      ],
    });

    let maxEmissionsPerSecond: string;
    if (maxEmissionsPerSecondChoice == 'units') {
      const tokenUnits = await numberPrompt({
        message: 'Enter the token units for maxEmissionsPerSecond',
        required: true,
      });
      const days = await numberPromptInDays({
        message: 'Enter the days for maxEmissionsPerSecond:',
        required: true,
      });
      const tokenDecimals = await getTokenDecimals(reward.asset, chainId);
      maxEmissionsPerSecond = `uint256(${tokenUnits} * 1e${tokenDecimals}) / ${days} days`;
    } else if (maxEmissionsPerSecondChoice == 'raw') {
      maxEmissionsPerSecond = await numberPromptNoTransform(
        {message: 'Enter the maxEmissionsPerSecond raw value', required: true},
        {
          skipTransform: true,
        }
      );
    } else {
      maxEmissionsPerSecond = 'EngineFlags.KEEP_CURRENT';
    }

    const distributionEndChoice = await select({
      message: `Please input the distributionEnd you want to configure for the reward: ${reward.symbol} and asset: ${asset}`,
      choices: [
        {name: 'Keep distributionEnd the same as current', value: 'current'},
        {name: 'Enter distributionEnd in days from current timestamp', value: 'units'},
        {name: 'Enter raw maxEmissionsPerSecond in unix timestamp', value: 'raw'},
      ],
    });

    let distributionEnd: string;
    if (distributionEndChoice == 'units') {
      distributionEnd = await numberPromptInDays({
        message: 'Enter the distributionEnd in days from current timestamp:',
        required: true,
      });
      distributionEnd = `block.timestamp + ${distributionEnd} days`;
    } else if (distributionEndChoice == 'raw') {
      distributionEnd = await numberPromptNoTransform(
        {message: 'Enter the distributionEnd in unix timestamp', required: true},
        {
          skipTransform: true,
        }
      );
    } else {
      distributionEnd = 'EngineFlags.KEEP_CURRENT';
    }

    let rewardPayer = await select({
      message: 'Enter the address of the rewards payer you want to configure',
      choices: [
        {name: 'Aave Collector (Default)', value: getDefaultCollector(pool)},
        {name: 'Custom Address (Enter Manually)', value: 'custom'},
      ],
    });
    if (rewardPayer == 'custom') {
      rewardPayer = await addressPrompt({
        message: 'Enter the address of the new rewards payer you want to configure:',
        required: true,
      });
    }

    input.push({
      asset,
      reward: reward.asset,
      rewardPayer,
      maxEmissionsPerSecond,
      distributionEnd,
    });
  }

  return input;
}

export const updateUmbrellaRewards: FeatureModule<UmbrellaRewardsUpdate[]> = {
  value: FEATURE.UPDATE_UMBRELLA_REWARDS,
  description: 'Updating existing Umbrella Rewards',
  async cli({pool}) {
    console.log(`Fetching information for updating existing umbrella rewards on ${pool}`);
    const response: UmbrellaRewardsUpdate[] = await fetchUmbrellaRewardsUpdateParams({pool});
    return response;
  },
  async build({pool, cfg}) {
    const rewardSymbols = await getTokenSymbols(
      cfg.map((s) => s.reward as Hex),
      CHAIN_TO_CHAIN_ID[getPoolChain(pool)]
    );
    const response: CodeArtifact = {
      code: {
        fn: [
          `
          function configureUpdates() public view override returns (RewardConfig[] memory) {
          RewardConfig[] memory newConfig = new RewardConfig[](${cfg.length});

          ${cfg
            .map(
              (cfg, ix) => `newConfig[${ix}] = RewardConfig({
               asset: ${translateUmbrellaAssetLibUnderlying(cfg.asset, pool)},
               reward: ${cfg.reward}, // ${rewardSymbols[ix]}
               rewardPayer: ${cfg.rewardPayer},
               maxEmissionPerSecond: ${cfg.maxEmissionsPerSecond},
               distributionEnd: ${cfg.distributionEnd}
             });`
            )
            .join('\n')}

          return newConfig;
        }
        `,
        ],
      },
    };
    return response;
  },
};
