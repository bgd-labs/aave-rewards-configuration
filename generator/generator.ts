import fs from 'fs';
import path from 'path';
import {generateContractName, generateFolderName, generateScriptName, getPoolChain} from './common';
import {liquidityMiningSetupTemplate} from './templates/liquiditymining.setup.template';
import {liquidityMiningUpdateTemplate} from './templates/liquiditymining.update.template';
import {liquidityMiningPayloadDeploymentTemplate} from './templates/liquiditymining.payloaddeployment.template';
import {confirm} from '@inquirer/prompts';
import {ConfigFile, Options, PoolConfigs, PoolIdentifier, Files} from './types';
import prettier from 'prettier';

const prettierSolCfg = await prettier.resolveConfig('foo.sol');
const prettierTsCfg = await prettier.resolveConfig('foo.ts');

/**
 * Generates all the file contents for aip/tests/payloads & script
 * @param options
 * @param poolConfigs
 * @returns
 */
export async function generateFiles(options: Options, poolConfigs: PoolConfigs): Promise<Files> {
  const jsonConfig = await prettier.format(
    `import {ConfigFile} from '../../generator/types';
    export const config: ConfigFile = ${JSON.stringify({
      rootOptions: options,
      poolOptions: (Object.keys(poolConfigs) as PoolIdentifier[]).reduce((acc, pool) => {
        acc[pool] = {configs: poolConfigs[pool]!.configs, cache: poolConfigs[pool]!.cache};
        return acc;
      }, {}),
    } as ConfigFile)}`,
    {...prettierTsCfg, filepath: 'foo.ts'}
  );

  const contractName = generateContractName(options, options.pool);
  async function createPayloadTest(options: Options, pool: PoolIdentifier) {
    const isLMSetup = options.feature == 'SETUP_LM';

    return {
      pool,
      payloadTest: await prettier.format(
        isLMSetup
          ? liquidityMiningSetupTemplate(options, poolConfigs[pool]!, pool)
          : liquidityMiningUpdateTemplate(options, poolConfigs[pool]!, pool),
        {
          ...prettierSolCfg,
          filepath: 'foo.sol',
        }
      ),
      contractName: contractName,
    };
  }

  async function createPayloadDeployment(options: Options, pool: PoolIdentifier) {
    const scriptName = generateScriptName(options, pool);
    const code = liquidityMiningPayloadDeploymentTemplate(options);
    return {
      payloadDeployment: await prettier.format(code, {
        ...prettierSolCfg,
        filepath: 'foo.sol',
      }),
      scriptName: scriptName,
    };
  }

  return {
    jsonConfig,
    payloadTest: await createPayloadTest(options, options.pool),
    payloadDeployment: await createPayloadDeployment(options, options.pool),
  };
}

async function askBeforeWrite(options: Options, path: string, content: string) {
  if (!options.force && fs.existsSync(path)) {
    const currentContent = fs.readFileSync(path, {encoding: 'utf8'});
    // skip if content did not change
    if (currentContent === content) return;
    const force = await confirm({
      message: `A file already exists at ${path} do you want to overwrite`,
      default: false,
    });
    if (!force) return;
  }
  fs.writeFileSync(path, content);
}

/**
 * Writes the files according to defined folder/file format
 */
export async function writeFiles(
  options: Options,
  {jsonConfig, payloadTest, payloadDeployment}: Files
) {
  const baseName = generateFolderName(options);
  const baseFolder = path.join(process.cwd(), 'tests', baseName);

  if (fs.existsSync(baseFolder)) {
    if (!options.force && fs.existsSync(baseFolder)) {
      const force = await confirm({
        message: 'A proposal already exists at that location, do you want to continue?',
        default: false,
      });
      if (!force) return;
    }
  } else {
    fs.mkdirSync(baseFolder, {recursive: true});
  }

  // write config
  await askBeforeWrite(options, path.join(baseFolder, 'config.ts'), jsonConfig);
  // write payloadTest
  await askBeforeWrite(
    options,
    path.join(baseFolder, `${payloadTest.contractName}.t.sol`),
    payloadTest.payloadTest
  );
  // write payloadDeployment
  await askBeforeWrite(
    options,
    path.join(baseFolder, `${payloadDeployment.scriptName}.s.sol`),
    payloadDeployment.payloadDeployment
  );
}
