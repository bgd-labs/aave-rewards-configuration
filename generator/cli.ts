import path from 'path';
import {Command, Option} from 'commander';
import {CHAIN_TO_CHAIN_ID, getDate, getPoolChain, pascalCase} from './common';
import {input, select} from '@inquirer/prompts';
import {ConfigFile, FEATURE, Options, POOLS, PoolCache, PoolConfigs, PoolIdentifier} from './types';
import {generateFiles, writeFiles} from './generator';
import {CHAIN_ID_CLIENT_MAP} from '@bgd-labs/js-utils';
import {getBlockNumber} from 'viem/actions';
import {setupLiquidityMining} from './features/setupLiquidityMining';
import {updateLiquidityMining} from './features/updateLiquidityMining';
import {updateUmbrellaRewards} from './features/updateUmbrellaRewards';
import {LiquidityMiningSetup, LiquidityMiningUpdate, UmbrellaRewardsUpdate} from './features/types';

const program = new Command();

program
  .name('aave-reward-configurations-generator')
  .description('CLI to generate scripts for updating umbrella rewards and LM on aave')
  .version('1.0.0')
  .addOption(new Option('-f, --force', 'force creation (might overwrite existing files)'))
  .addOption(new Option('-c, --configFile <string>', 'path to config file'))
  .allowExcessArguments(false)
  .parse(process.argv);

let options = program.opts<Options>();
let poolConfigs: PoolConfigs = {};

const FEATURE_MODULES = [setupLiquidityMining, updateLiquidityMining, updateUmbrellaRewards];

async function generateDeterministicPoolCache(pool: PoolIdentifier): Promise<PoolCache> {
  const chain = getPoolChain(pool);
  const client = CHAIN_ID_CLIENT_MAP[CHAIN_TO_CHAIN_ID[chain]];
  return {blockNumber: Number(await getBlockNumber(client))};
}

async function fetchLMSetupOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const feature = FEATURE.SETUP_LM;
  const module = setupLiquidityMining;
  poolConfigs[pool]!.configs[feature] = await module.cli({
    options,
    pool,
    cache: poolConfigs[pool]!.cache,
  });
  poolConfigs[pool]!.artifacts.push(
    await module.build({
      options,
      pool,
      cfg: poolConfigs[pool]!.configs[feature],
      cache: poolConfigs[pool]!.cache,
    })
  );
}

async function fetchLMUpdateOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const feature = FEATURE.UPDATE_LM;
  const module = updateLiquidityMining;
  poolConfigs[pool]!.configs[feature] = await module.cli({
    options,
    pool,
    cache: poolConfigs[pool]!.cache,
  });
  poolConfigs[pool]!.artifacts.push(
    await module.build({
      options,
      pool,
      cfg: poolConfigs[pool]!.configs[feature],
      cache: poolConfigs[pool]!.cache,
    })
  );
}

async function fetchUmbrellaRewardsOptions(pool: PoolIdentifier) {
  poolConfigs[pool] = {
    configs: {},
    artifacts: [],
    cache: await generateDeterministicPoolCache(pool),
  };

  const feature = FEATURE.UPDATE_UMBRELLA_REWARDS;
  const module = updateUmbrellaRewards;
  poolConfigs[pool]!.configs[feature] = await module.cli({
    options,
    pool,
    cache: poolConfigs[pool]!.cache,
  });
  poolConfigs[pool]!.artifacts.push(
    await module.build({
      options,
      pool,
      cfg: poolConfigs[pool]!.configs[feature],
      cache: poolConfigs[pool]!.cache,
    })
  );
}

if (options.configFile) {
  const {config: cfgFile}: {config: ConfigFile} = await import(
    path.join(process.cwd(), options.configFile)
  );
  options = {...options, ...cfgFile.rootOptions};
  poolConfigs = cfgFile.poolOptions as any;
  poolConfigs[options.pool]!.artifacts = [];

  if (options.feature == FEATURE.SETUP_LM) {
    poolConfigs[options.pool]!.artifacts.push(
      await setupLiquidityMining.build({
        options,
        pool: options.pool,
        cfg: poolConfigs[options.pool]!.configs[FEATURE.SETUP_LM] as LiquidityMiningSetup,
        cache: poolConfigs[options.pool]!.cache,
      })
    );
  } else if (options.feature == FEATURE.UPDATE_LM) {
    poolConfigs[options.pool]!.artifacts.push(
      await updateLiquidityMining.build({
        options,
        pool: options.pool,
        cfg: poolConfigs[options.pool]!.configs[FEATURE.UPDATE_LM] as LiquidityMiningUpdate,
        cache: poolConfigs[options.pool]!.cache,
      })
    );
  } else {
    poolConfigs[options.pool]!.artifacts.push(
      await updateUmbrellaRewards.build({
        options,
        pool: options.pool,
        cfg: poolConfigs[options.pool]!.configs[
          FEATURE.UPDATE_UMBRELLA_REWARDS
        ] as UmbrellaRewardsUpdate[],
        cache: poolConfigs[options.pool]!.cache,
      })
    );
  }
} else {
  options.feature = await select({
    message: 'Please select the reward action you want to do',
    choices: FEATURE_MODULES.map((m) => ({name: m.description, value: m.value})),
  });

  options.pool = await select({
    message: 'Select the Aave Pool:',
    choices:
      options.feature === FEATURE.UPDATE_UMBRELLA_REWARDS
        ? [
            {name: 'AaveV3Ethereum', value: 'AaveV3Ethereum'},
            {name: 'AaveV3BaseSepolia', value: 'AaveV3BaseSepolia'},
          ]
        : POOLS.map((v) => ({name: v, value: v})),
  });

  if (!options.title) {
    options.title = await input({
      message: 'Short title of the rewards program:',
      validate(input) {
        if (input.length == 0) return "Your title can't be empty";
        if (input.trim().length > 80) return 'Your title is to long';
        return true;
      },
    });
  }
  options.shortName = pascalCase(options.title);
  options.date = getDate();

  if (options.feature === FEATURE.SETUP_LM) {
    await fetchLMSetupOptions(options.pool);
  } else if (options.feature === FEATURE.UPDATE_LM) {
    await fetchLMUpdateOptions(options.pool);
  } else {
    await fetchUmbrellaRewardsOptions(options.pool);
  }
}

try {
  const files = await generateFiles(options, poolConfigs);
  await writeFiles(options, files);
} catch (e) {
  console.log(JSON.stringify({options, poolConfigs}, null, 2));
  throw e;
}
