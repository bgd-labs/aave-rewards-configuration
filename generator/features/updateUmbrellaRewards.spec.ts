import {expect, describe, it} from 'vitest';
import {MOCK_OPTIONS_UMBRELLA, umbrellaRewardsUpdateConfig} from './mocks/configs';
import {generateFiles} from '../generator';
import {FEATURE, PoolConfigs} from '../types';
import {updateUmbrellaRewards} from './updateUmbrellaRewards';

describe('feature: updateUmbrellaRewards', () => {
  it('should return reasonable code', async () => {
    const output = await updateUmbrellaRewards.build({
      options: MOCK_OPTIONS_UMBRELLA,
      pool: MOCK_OPTIONS_UMBRELLA.pool,
      cfg: umbrellaRewardsUpdateConfig,
      cache: {blockNumber: 42},
    });
    expect(output).toMatchSnapshot();
  });

  it('should properly generate files', async () => {
    const poolConfigs: PoolConfigs = {
      [MOCK_OPTIONS_UMBRELLA.pool]: {
        pool: MOCK_OPTIONS_UMBRELLA.pool,
        artifacts: [
          await updateUmbrellaRewards.build({
            options: MOCK_OPTIONS_UMBRELLA,
            pool: MOCK_OPTIONS_UMBRELLA.pool,
            cfg: umbrellaRewardsUpdateConfig,
            cache: {blockNumber: 42},
          }),
        ],
        configs: {[FEATURE.UPDATE_LM]: umbrellaRewardsUpdateConfig},
        cache: {blockNumber: 42},
      },
    };
    const files = await generateFiles(MOCK_OPTIONS_UMBRELLA, poolConfigs);
    expect(files).toMatchSnapshot();
  });
});
