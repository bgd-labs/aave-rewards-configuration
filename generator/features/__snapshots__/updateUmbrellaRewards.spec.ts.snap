// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`feature: updateUmbrellaRewards > should properly generate files 1`] = `
{
  "jsonConfig": "import {ConfigFile} from '../../generator/types';
export const config: ConfigFile = {
  rootOptions: {
    pool: 'AaveV3BaseSepolia',
    title: 'test',
    shortName: 'Test',
    date: '20250602',
    feature: 'UPDATE_UMBRELLA_REWARDS',
  },
  poolOptions: {
    AaveV3BaseSepolia: {
      configs: {
        UPDATE_LM: [
          {
            asset: 'STK_WETH_V1',
            reward: '0x73a5bB60b0B0fc35710DDc0ea9c407031E31Bdbb',
            rewardPayer: '0x24f321Cd3742B7a191e5ED7cE1be8059DEe91424',
            maxEmissionsPerSecond: 'EngineFlags.KEEP_CURRENT',
            distributionEnd: 'block.timestamp + 120 days',
          },
        ],
      },
      cache: {blockNumber: 42},
    },
  },
};
",
  "payloadTest": {
    "contractName": "AaveV3BaseSepolia_UmbrellaRewardsUpdateTest_20250602",
    "payloadTest": "// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import {EngineFlags} from 'aave-v3-origin/contracts/extensions/v3-config-engine/EngineFlags.sol';
import {UmbrellaBaseSepoliaAssets} from 'aave-address-book/UmbrellaBaseSepolia.sol';
import {UmbrellaRewardsBaseTest} from '../utils/UmbrellaRewardsBaseTest.sol';
import {UmbrellaBaseSepoliaConfig} from '../utils/networkConfig/UmbrellaBaseSepoliaConfig.sol';

/**
 * private-key: forge test --mp tests/20250602_UmbrellaUpdateAaveV3BaseSepolia_Test/AaveV3BaseSepolia_UmbrellaRewardsUpdateTest_20250602.t.sol --mt test_sendTransactionViaPrivateKey -vv
 * ledger: forge test --mp tests/20250602_UmbrellaUpdateAaveV3BaseSepolia_Test/AaveV3BaseSepolia_UmbrellaRewardsUpdateTest_20250602.t.sol --mt test_sendTransactionViaLedger -vv
 * emit Calldata only: forge test --mp tests/20250602_UmbrellaUpdateAaveV3BaseSepolia_Test/AaveV3BaseSepolia_UmbrellaRewardsUpdateTest_20250602.t.sol --mt test_logCalldatas -vv
 */
contract AaveV3BaseSepolia_UmbrellaRewardsUpdateTest_20250602 is
  UmbrellaRewardsBaseTest,
  UmbrellaBaseSepoliaConfig
{
  function configureUpdates() public pure override returns (RewardConfig[] memory) {
    RewardConfig[] memory newConfig = new RewardConfig[](1);

    newConfig[0] = RewardConfig({
      asset: UmbrellaBaseSepoliaAssets.STK_WETH_V1,
      reward: 0x73a5bB60b0B0fc35710DDc0ea9c407031E31Bdbb, // aBasSepWETH
      rewardPayer: 0x24f321Cd3742B7a191e5ED7cE1be8059DEe91424,
      maxEmissionPerSecond: EngineFlags.KEEP_CURRENT,
      distributionEnd: block.timestamp + 120 days
    });

    return newConfig;
  }
}
",
    "pool": "AaveV3BaseSepolia",
  },
}
`;

exports[`feature: updateUmbrellaRewards > should return reasonable code 1`] = `
{
  "code": {
    "fn": [
      "
          function configureUpdates() public pure override returns (RewardConfig[] memory) {
          RewardConfig[] memory newConfig = new RewardConfig[](1);

          newConfig[0] = RewardConfig({
               asset: UmbrellaBaseSepoliaAssets.STK_WETH_V1,
               reward: 0x73a5bB60b0B0fc35710DDc0ea9c407031E31Bdbb, // aBasSepWETH
               rewardPayer: 0x24f321Cd3742B7a191e5ED7cE1be8059DEe91424,
               maxEmissionPerSecond: EngineFlags.KEEP_CURRENT,
               distributionEnd: block.timestamp + 120 days
             });

          return newConfig;
        }
        ",
    ],
  },
}
`;
